src/convert.jl

```diff
@@ -1,14 +1,150 @@
-(::Type{SA})(x::Tuple{Tuple{Tuple{<:Tuple}}}) where {SA <: StaticArray} =
-    throw(DimensionMismatch("No precise constructor for $SA found. Length of input was $(length(x[1][1][1]))."))
+# A help wrapper to distinguish `SA(x...)` and `SA((x...，))`
+struct Args{T<:Tuple}
+    args::T
+end
+Length(x::Args) = Length(length(x.args))
+const BadArgs = Args{<:Tuple{Tuple{<:Tuple}}}
+
+# Some help functions.
+@pure has_ndims(::Type{<:StaticArray{<:Tuple,<:Any,N}}) where {N} = @isdefined N
+@pure has_ndims(::Type{<:StaticArray}) = false
+if VERSION < v"1.7"
+    Base.ndims(::Type{<:StaticArray{<:Tuple,<:Any,N}}) where {N} = N
+end
+@pure has_eltype(::Type{<:StaticArray{<:Tuple,T}}) where {T} = @isdefined T
+@pure has_eltype(::Type{<:StaticArray}) = false
+@pure has_size(::Type{<:StaticArray{S}}) where {S<:Tuple} = @isdefined S
+@pure has_size(::Type{<:StaticArray}) = false
+@pure has_size1(::Type{<:StaticMatrix{M}}) where {M} = @isdefined M
+@pure has_size1(::Type{<:StaticMatrix}) = false
+_size1(::Type{<:StaticMatrix{M}}) where {M} = M
+StaticSquareMatrix{N,T} = StaticMatrix{N,N,T}
+@generated function _sqrt(::Length{L}) where {L}
+    N = round(Int, sqrt(L))
+    N^2 == L || throw(DimensionMismatch("Input's length must be perfect square"))
+    return :($N)
+end
+
+const FirstClass = Union{SArray, MArray, SHermitianCompact, SizedArray}
```

Comment:
What's the meaning of `FirstClass`? Should this include `FieldArray`? I thought a few more comments would make it easier to review:+1:

Reply:
I have renamed it as `SizeEltypeAdaptable` (`FieldArray` only do eltype adaption) Not sure is it a better name though.